src/logic/tokens/store/actions/fetchSafeTokens.ts

```diff
@@ -1,67 +1,92 @@
+const extractDataFromResult = (currentTokens: TokenState) => (
+  acc,
+  { balance, balanceUsd, token, tokenAddress },
+): ExtractedData => {
+  if (tokenAddress === null) {
+    acc.ethBalance = humanReadableValue(balance, 18)
+  } else {
+    acc.balances = acc.balances.merge({ [tokenAddress]: humanReadableValue(balance, token.decimals) })
+
+    if (currentTokens && !currentTokens.get(tokenAddress)) {
+      acc.tokens = acc.tokens.push(makeToken({ address: tokenAddress, ...token }))
+    }
+  }
+
+  acc.currencyList = acc.currencyList.push(
+    makeBalanceCurrency({
+      currencyName: balanceUsd ? AVAILABLE_CURRENCIES.USD : null,
+      tokenAddress,
+      balanceInBaseCurrency: balanceUsd,
+      balanceInSelectedCurrency: balanceUsd,
+    }),
+  )
+
+  return acc
+}
 
 const fetchSafeTokens = (safeAddress: string) => async (
   dispatch: Dispatch,
   getState: () => AppReduxState,
 ): Promise<void> => {
   try {
     const state = getState()
-    const safe = state[SAFE_REDUCER_ID].getIn(['safes', safeAddress])
-    const currentTokens = state[TOKEN_REDUCER_ID]
+    const safe = safeSelector(state)
+    const currentTokens = tokensSelector(state)
 
     if (!safe) {
       return
     }
 
     const result = await backOff(() => fetchTokenCurrenciesBalances(safeAddress))
-    const currentEthBalance = safe.get('ethBalance')
-    const safeBalances = safe.get('balances')
-    const alreadyActiveTokens = safe.get('activeTokens')
-    const blacklistedTokens = safe.get('blacklistedTokens')
-    const currencyValues = state[CURRENCY_VALUES_KEY]
-
-    const { balances, currencyList, ethBalance, tokens } = result.data.reduce(
-      (acc, { balance, balanceUsd, token, tokenAddress }) => {
-        if (tokenAddress === null) {
-          acc.ethBalance = humanReadableBalance(balance, 18)
-        } else {
-          acc.balances = acc.balances.merge({ [tokenAddress]: humanReadableBalance(balance, token.decimals) })
-
-          if (currentTokens && !currentTokens.get(tokenAddress)) {
-            acc.tokens = acc.tokens.push(makeToken({ address: tokenAddress, ...token }))
-          }
-        }
-
-        acc.currencyList = acc.currencyList.push(
-          makeBalanceCurrency({
-            currencyName: balanceUsd ? AVAILABLE_CURRENCIES.USD : null,
-            tokenAddress,
-            balanceInBaseCurrency: balanceUsd,
-            balanceInSelectedCurrency: balanceUsd,
-          }),
-        )
-
-        return acc
-      },
+    const currentEthBalance = safeEthBalanceSelector(state)
+    const safeBalances = safeBalancesSelector(state)
+    const alreadyActiveTokens = safeActiveTokensSelector(state)
+    const blacklistedTokens = safeBlacklistedTokensSelector(state)
+    const currencyValues = currencyValuesSelector(state)
+
+    const { balances, currencyList, ethBalance, tokens }: ExtractedData = result.data.reduce(
+      extractDataFromResult(currentTokens),
```

Comment:
https://stackoverflow.com/questions/28510753/when-should-i-use-withmutations-on-a-map-in-immutable-js

Reply:
Ok, once again. This change was just moving code around for better readability. We should address this in a different issue

Reply:
could you please create one?

Reply:
#1184